device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/MultiplexingClient.java

```diff
@@ -0,0 +1,257 @@
+     */
+    public void registerDeviceClient(DeviceClient deviceClient)
+    {
+        Objects.requireNonNull(deviceClient);
+
+        synchronized (this.lock)
+        {
+            for (DeviceClient currentClient : deviceClientList)
+            {
+                String currentDeviceId = currentClient.getConfig().getDeviceId();
+                if (currentDeviceId.equalsIgnoreCase(deviceClient.getConfig().getDeviceId()))
+                {
+                    log.warn("Device {} won't be registered to the multiplexed connection because it is already registered.", currentDeviceId);
+                    return;
+                }
+            }
+
+            if (deviceClient.getConfig().getAuthenticationType() != DeviceClientConfig.AuthType.SAS_TOKEN)
+            {
+                throw new UnsupportedOperationException("Can only register a device client if it uses SAS token based authentication");
+            }
+
+            if (deviceClient.getConfig().getProtocol() != this.protocol)
+            {
+                throw new UnsupportedOperationException("A device client cannot be registered to a multiplexing client that specifies a different transport protocol.");
+            }
+
+            if (this.protocol == IotHubClientProtocol.AMQPS && this.deviceClientList.size() > MAX_MULTIPLEX_DEVICE_COUNT_AMQPS)
+            {
+                throw new UnsupportedOperationException(String.format("Multiplexed connections over AMQPS only support up to %d devices", MAX_MULTIPLEX_DEVICE_COUNT_AMQPS));
+            }
+
+            // Typically client side validation is duplicate work, but IoT Hub doesn't give a good error message when closing the
+            // AMQPS_WS connection so this is the only way that users will know about this limit
+            if (this.protocol == IotHubClientProtocol.AMQPS_WS && this.deviceClientList.size() > MAX_MULTIPLEX_DEVICE_COUNT_AMQPS_WS)
+            {
+                throw new UnsupportedOperationException(String.format("Multiplexed connections over AMQPS_WS only support up to %d devices", MAX_MULTIPLEX_DEVICE_COUNT_AMQPS_WS));
+            }
+
+            if (this.deviceClientList.size() > 1 && !this.deviceClientList.get(0).getConfig().getIotHubHostname().equalsIgnoreCase(deviceClient.getConfig().getIotHubHostname()))
+            {
+                throw new UnsupportedOperationException("Multiplexed device clients must connect to the same host name");
+            }
+
+            // The first device to be registered will cause this client to build the IO layer with its configuration
+            if (this.deviceIO == null)
+            {
+                log.debug("Creating DeviceIO layer for multiplexing client since this is the first registered device");
+                this.deviceIO = new DeviceIO(deviceClient.getConfig(), SEND_PERIOD_MILLIS, RECEIVE_PERIOD_MILLIS);
+            }
+
+            deviceClient.setDeviceIO(this.deviceIO);
```

Comment:
what if the deviceIO already exists on this device? (it has already been opened independently from the multiplex connection) 
what I mean is that the pre-existing device IO needs to be disposed. we also need tests for that :D

Reply:
I think I'll throw in that scenario. To me, the scenario where a user tries to register a device client instance that is already open sounds more accidental than deliberate